File: harness/oracle/cases/context-summary.ts
/**
 * Oracle case #17 (Wave 3 / B8): context-summary gating cadence.
 *
 * Drives the REAL pure functions from v4's lib/chat/context-summary.ts:
 *   evaluateSummarizationGate, calculateInterchangeCount,
 *   shouldCheckTitleAtInterchange, partitionMessagesIntoTurns.
 * All four are side-effect-free; no injection needed.
 *
 * Run from inside the server checkout:
 *   cd ~/source/quilltap-server
 *   npx tsx ~/source/quilltap-v5/harness/oracle/cases/context-summary.ts \
 *     > /tmp/oracle-context-summary.ndjson
 */

import {
  evaluateSummarizationGate,
  calculateInterchangeCount,
  shouldCheckTitleAtInterchange,
  partitionMessagesIntoTurns,
} from '@/lib/chat/context-summary';
import type { ChatEvent } from '@/lib/schemas/types';

type Row =
  | { kind: 'gate'; id: string; currentTurn: number; lastFoldedTurn: number; lastFullRebuildTurn: number; out: string }
  | { kind: 'interchange'; id: string; messages: Array<{ role?: string; type?: string; systemSender?: string | null }>; chatType?: string; out: number }
  | { kind: 'title'; id: string; current: number; lastChecked: number; chatType?: string; out: boolean }
  | { kind: 'partition'; id: string; messages: Array<Record<string, unknown>>; chatType?: string; out: Array<{ turnNumber: number; ids: string[] }> };

const rows: Row[] = [];

// ---- evaluateSummarizationGate -------------------------------------------
const gates: Array<[string, number, number, number]> = [
  ['below-floor-0', 0, 0, 0],
  ['below-floor-5', 5, 0, 0],
  ['at-floor-10', 10, 0, 0],
  ['just-over-fold', 11, 0, 0],
  ['fold-12-1', 12, 1, 0],
  ['skip-12-2', 12, 2, 0],
  ['hard-60', 60, 0, 0],
  ['hard-boundary-61', 61, 5, 11],
  ['fold-not-hard', 60, 40, 20],
  ['skip-recent-fold', 60, 55, 20],
  ['fold-30', 30, 15, 25],
  ['skip-22', 22, 15, 10],
];
for (const [id, c, lf, lr] of gates) {
  rows.push({ kind: 'gate', id, currentTurn: c, lastFoldedTurn: lf, lastFullRebuildTurn: lr, out: evaluateSummarizationGate({ currentTurn: c, lastFoldedTurn: lf, lastFullRebuildTurn: lr }) });
}

// ---- calculateInterchangeCount -------------------------------------------
const mU = { role: 'USER', type: 'message' };
const mA = { role: 'ASSISTANT', type: 'message' };
const mWhisper = { role: 'ASSISTANT', type: 'message', systemSender: 'librarian' };
const interchanges: Array<[string, Array<{ role?: string; type?: string; systemSender?: string | null }>, string | undefined]> = [
  ['empty', [], undefined],
  ['one-pair', [mU, mA], undefined],
  ['two-user-one-assistant', [mU, mU, mA], undefined],
  ['one-user-two-assistant', [mU, mA, mA], undefined],
  ['whisper-excluded', [mU, mWhisper, mA], undefined],
  ['nonmessage-type-excluded', [mU, { role: 'ASSISTANT', type: 'context-summary' }, mA], undefined],
  ['lowercase-role', [{ role: 'user', type: 'message' }, { role: 'assistant', type: 'message' }], undefined],
  ['empty-type-is-message', [{ role: 'USER', type: '' }, { role: 'ASSISTANT', type: '' }], undefined],
  ['undefined-role-skipped', [{ type: 'message' }, mU, mA], undefined],
  ['autonomous-counts-assistant', [mA, mA, mA], 'autonomous'],
  ['autonomous-whisper-excluded', [mA, mWhisper, mA, mWhisper, mA], 'autonomous'],
  ['autonomous-no-user-floor', [mU, mA, mA], 'autonomous'],
  ['empty-systemSender-counts', [mU, { role: 'ASSISTANT', type: 'message', systemSender: '' }], undefined],
];
for (const [id, messages, chatType] of interchanges) {
  rows.push({ kind: 'interchange', id, messages, chatType, out: calculateInterchangeCount(messages, chatType) });
}

// ---- shouldCheckTitleAtInterchange ---------------------------------------
const titles: Array<[string, number, number, string | undefined]> = [
  ['reg-below-min', 1, 0, undefined],
  ['reg-checkpoint-2', 2, 0, undefined],
  ['reg-already-2', 2, 2, undefined],
  ['reg-gap-4', 4, 3, undefined],
  ['reg-checkpoint-5', 5, 3, undefined],
  ['reg-12-after-10', 12, 10, undefined],
  ['reg-20-after-10', 20, 10, undefined],
  ['reg-25-after-19', 25, 19, undefined],
  ['reg-jump-23', 23, 8, undefined],
  ['help-checkpoint-1', 1, 0, 'help'],
  ['help-already-1', 1, 1, 'help'],
  ['help-brahma-1', 1, 0, 'brahma'],
  ['autonomous-as-regular', 15, 5, 'autonomous'],
];
for (const [id, current, lastChecked, chatType] of titles) {
  rows.push({ kind: 'title', id, current, lastChecked, chatType, out: shouldCheckTitleAtInterchange(current, lastChecked, chatType) });
}

// ---- partitionMessagesIntoTurns ------------------------------------------
const ev = (id: string, role: string, type = 'message', systemSender: string | null = null): Record<string, unknown> =>
  ({ id, role, type, content: `c-${id}`, systemSender });
const partitions: Array<[string, Array<Record<string, unknown>>, string | undefined]> = [
  ['empty', [], undefined],
  ['two-turns', [ev('u1', 'USER'), ev('a1', 'ASSISTANT'), ev('u2', 'USER'), ev('a2', 'ASSISTANT')], undefined],
  ['leading-greeting', [ev('a0', 'ASSISTANT'), ev('u1', 'USER'), ev('a1', 'ASSISTANT')], undefined],
  ['staff-whisper-excluded', [ev('u1', 'USER'), ev('w1', 'ASSISTANT', 'message', 'librarian'), ev('a1', 'ASSISTANT')], undefined],
  ['nonmessage-skipped', [ev('u1', 'USER'), ev('cs', 'ASSISTANT', 'context-summary'), ev('a1', 'ASSISTANT')], undefined],
  ['assistant-only-no-flush', [ev('a1', 'ASSISTANT'), ev('a2', 'ASSISTANT')], undefined],
  ['autonomous-each-assistant', [ev('a1', 'ASSISTANT'), ev('a2', 'ASSISTANT'), ev('a3', 'ASSISTANT')], 'autonomous'],
  ['autonomous-whisper-excluded', [ev('a1', 'ASSISTANT'), ev('w1', 'ASSISTANT', 'message', 'librarian'), ev('a2', 'ASSISTANT')], 'autonomous'],
];
for (const [id, messages, chatType] of partitions) {
  const turns = partitionMessagesIntoTurns(messages as unknown as ChatEvent[], chatType);
  rows.push({ kind: 'partition', id, messages, chatType, out: turns.map(t => ({ turnNumber: t.turnNumber, ids: t.ids })) });
}

for (const r of rows) process.stdout.write(JSON.stringify(r) + '\n');

File: crates/quilltap-core/src/context_summary.rs
//! Port of the pure gating logic in v4's lib/chat/context-summary.ts — the
//! rolling-window summarisation cadence. These four functions decide *when* a
//! fold / rebuild / title-check fires; the actual LLM fold, the DB writes and
//! the Librarian-whisper sweep stay in later (stateful) phases.
//!
//! Cadence knobs (carried verbatim from v4):
//!   - FOLD_TURN_BATCH: turns folded per fire (5).
//!   - FOLD_TAIL_FLOOR: minimum recent turns kept verbatim (5).
//!   - FOLD_TRIGGER_DELTA: accumulated tail before a fold fires (floor + batch
//!     = 10). When `currentTurn - lastFoldedTurn > FOLD_TRIGGER_DELTA`, fold.
//!   - T_HARD_TURN_THRESHOLD: periodic full from-scratch rebuild (50), cheap
//!     insurance against accumulated paraphrase drift across many folds.

use crate::chat_predicates::is_help_like_chat_type;

/// Number of turns folded per fire.
pub const FOLD_TURN_BATCH: i64 = 5;
/// Minimum recent turns kept verbatim (never folded away).
pub const FOLD_TAIL_FLOOR: i64 = 5;
/// Turns of accumulated tail before a fold fires.
pub const FOLD_TRIGGER_DELTA: i64 = FOLD_TAIL_FLOOR + FOLD_TURN_BATCH;
/// Periodic full from-scratch rebuild threshold.
pub const T_HARD_TURN_THRESHOLD: i64 = 50;

/// The gate's verdict: leave the summary alone, fold the next batch, or do a
/// full from-scratch rebuild.
#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum SummarizationGateDecision {
    Skip,
    Fold,
    Hard,
}

impl SummarizationGateDecision {
    pub fn as_str(self) -> &'static str {
        match self {
            SummarizationGateDecision::Skip => "skip",
            SummarizationGateDecision::Fold => "fold",
            SummarizationGateDecision::Hard => "hard",
        }
    }
}

/// Decide whether to skip, fold the next batch, or do a full rebuild. Pure —
/// no side effects. `current_turn` is the interchange count; `last_folded_turn`
/// is the turn through which the running summary has already absorbed content;
/// `last_full_rebuild_turn` anchors the T_hard window.
pub fn evaluate_summarization_gate(
    current_turn: i64,
    last_folded_turn: i64,
    last_full_rebuild_turn: i64,
) -> SummarizationGateDecision {
    // Below the floor + batch threshold, no fold needs to happen — the LLM
    // still sees the full conversation as recent tail.
    if current_turn <= FOLD_TRIGGER_DELTA {
        return SummarizationGateDecision::Skip;
    }

    // T_hard wins over the regular fold path. A from-scratch rebuild implicitly
    // satisfies any fold that was due.
    if current_turn - last_full_rebuild_turn >= T_HARD_TURN_THRESHOLD {
        return SummarizationGateDecision::Hard;
    }

    if current_turn - last_folded_turn > FOLD_TRIGGER_DELTA {
        return SummarizationGateDecision::Fold;
    }

    SummarizationGateDecision::Skip
}

/// A minimal view of a chat message for the cadence functions: only the three
/// fields they read. `role`/`message_type` are `Option` to preserve the JS
/// distinction between absent and empty (the two guards treat them
/// differently — see the call sites).
#[derive(Clone, Debug)]
pub struct InterchangeMessage {
    pub role: Option<String>,
    pub message_type: Option<String>,
    pub system_sender: Option<String>,
}

/// Count the interchanges in a chat.
///
/// For normal chats an interchange is one user message + one assistant
/// response — count is the minimum of the two so partial pairs don't tick the
/// meter. Autonomous rooms have no human user, so that floor is permanently
/// zero; instead, count each assistant turn as one interchange.
///
/// Staff whispers (`systemSender` set) carry role ASSISTANT but are synthetic
/// system messages, not real turns — excluded so the cadence isn't inflated by
/// the ~5 whispers each autonomous turn drags along.
pub fn calculate_interchange_count(
    messages: &[InterchangeMessage],
    chat_type: Option<&str>,
) -> i64 {
    let mut user_messages = 0i64;
    let mut assistant_messages = 0i64;

    for msg in messages {
        // Skip non-message types (context-summary, tool-result, etc.). JS:
        // `if (msg.type && msg.type !== 'message')` — an absent or empty type
        // is falsy and is treated as a real message.
        if msg
            .message_type
            .as_deref()
            .is_some_and(|t| !t.is_empty() && t != "message")
        {
            continue;
        }

        // Skip staff whispers — not interchange-bearing turns. Empty string is
        // falsy in JS, so only a non-empty systemSender skips.
        if msg.system_sender.as_deref().is_some_and(|s| !s.is_empty()) {
            continue;
        }

        // `msg.role?.toUpperCase()`.
        let role = msg.role.as_deref().map(|r| r.to_uppercase());
        match role.as_deref() {
            Some("USER") => user_messages += 1,
            Some("ASSISTANT") => assistant_messages += 1,
            _ => {}
        }
    }

    if chat_type == Some("autonomous") {
        return assistant_messages;
    }

    // An interchange is complete when both user and assistant have spoken —
    // the minimum of the two is the limiting factor.
    user_messages.min(assistant_messages)
}

/// Whether a title-update check should fire at this interchange count.
///
/// - Regular chats: 2, 3, 5, 7, 10, then every 10 after.
/// - Help-like chats: 1, 2, 3, 5, 7, 10, then every 10 after (fires right after
///   the first Q&A).
///
/// Crossing semantics: each checkpoint fires when the counter has *reached or
/// passed* it since the last check — autonomous rooms can bump the counter by
/// 5+ in one turn and skip past exact 10/20/30 marks, so `>=` (not `==`) is
/// required.
pub fn should_check_title_at_interchange(
    current_interchange: i64,
    last_checked_interchange: i64,
    chat_type: Option<&str>,
) -> bool {
    let is_help_chat = is_help_like_chat_type(chat_type);

    // Help-like chats fire at interchange 1; regular chats never before 2.
    let minimum_interchange = if is_help_chat { 1 } else { 2 };
    if current_interchange < minimum_interchange {
        return false;
    }

    // Fire if we've crossed any not-yet-checked early checkpoint.
    let early_checkpoints: &[i64] = if is_help_chat {
        &[1, 2, 3, 5, 7, 10]
    } else {
        &[2, 3, 5, 7, 10]
    };
    for &checkpoint in early_checkpoints {
        if current_interchange >= checkpoint && last_checked_interchange < checkpoint {
            return true;
        }
    }

    // After 10, fire if the most recently crossed multiple of 10 is one we
    // haven't checked yet. `floor(n / 10) * 10` is the highest multiple of 10
    // ≤ n; if that exceeds `lastCheckedInterchange`, a new boundary was crossed.
    if current_interchange >= 10 {
        let last_crossed_multiple_of_10 = (current_interchange / 10) * 10;
        if last_crossed_multiple_of_10 > last_checked_interchange {
            return true;
        }
    }

    false
}

/// A message as seen by [`partition_messages_into_turns`].
#[derive(Clone, Debug)]
pub struct PartitionInputMessage {
    pub id: String,
    pub role: Option<String>,
    pub message_type: Option<String>,
    pub system_sender: Option<String>,
}

/// One turn produced by [`partition_messages_into_turns`]: a 1-indexed number
/// and the IDs of the USER + non-staff ASSISTANT messages composing it,
/// chronological.
#[derive(Clone, Debug, PartialEq, Eq)]
pub struct FoldedTurn {
    pub turn_number: usize,
    pub ids: Vec<String>,
}

/// Walk chat history in chronological order, grouping messages into turns. A
/// turn begins on a USER message; trailing non-staff ASSISTANT messages before
/// the next USER message belong to that turn. Staff-authored messages
/// (`systemSender` set) are excluded but do not affect numbering. ASSISTANT-only
/// greeting messages before any USER message are folded into turn 1.
///
/// Autonomous rooms have no USER pivot — partition on each character-attributed
/// ASSISTANT message instead, else `turns.length` is permanently 0 and
/// summarisation always bails with "No messages to summarize".
///
/// NB v4 compares `m.role` against the uppercase literals without uppercasing
/// (MessageEvent.role is a `'USER' | 'ASSISTANT'` union) — mirrored here as a
/// case-sensitive compare, unlike [`calculate_interchange_count`].
pub fn partition_messages_into_turns(
    all_messages: &[PartitionInputMessage],
    chat_type: Option<&str>,
) -> Vec<FoldedTurn> {
    let mut turns: Vec<FoldedTurn> = Vec::new();
    // Index into `turns` of the in-progress turn, mirroring v4's `currentTurn`.
    let mut current_turn: Option<usize> = None;
    let mut leading_assistant: Option<Vec<String>> = None;
    let is_autonomous = chat_type == Some("autonomous");

    for msg in all_messages {
        if msg.message_type.as_deref() != Some("message") {
            continue;
        }
        let role = msg.role.as_deref();
        if role != Some("USER") && role != Some("ASSISTANT") {
            continue;
        }
        if msg.system_sender.as_deref().is_some_and(|s| !s.is_empty()) {
            continue;
        }

        let starts_new_turn = role == Some("USER") || (is_autonomous && role == Some("ASSISTANT"));

        if starts_new_turn {
            let turn_number = turns.len() + 1;
            let mut start_ids = leading_assistant.take().unwrap_or_default();
            start_ids.push(msg.id.clone());
            turns.push(FoldedTurn {
                turn_number,
                ids: start_ids,
            });
            current_turn = Some(turns.len() - 1);
        } else if let Some(idx) = current_turn {
            turns[idx].ids.push(msg.id.clone());
        } else {
            leading_assistant
                .get_or_insert_with(Vec::new)
                .push(msg.id.clone());
        }
    }

    turns
}
